crates/matrix-sdk/tests/integration/encryption/verification.rs

```diff
@@ -202,3 +235,117 @@ async fn test_own_verification() {
+        .build()
+        .await
+        .unwrap();
+
+    let bob_user_id = owned_user_id!("@bob:example.org");
+    let bob_device_id = owned_device_id!("B0B0B0B0B");
+    bob.restore_session(MatrixSession {
+        meta: SessionMeta { user_id: bob_user_id.clone(), device_id: bob_device_id.clone() },
+        tokens: MatrixSessionTokens { access_token: "1234".to_owned(), refresh_token: None },
+    })
+    .await
+    .unwrap();
+
+    server.add_known_device(&alice_device_id);
+    server.add_known_device(&bob_device_id);
+
+    let alice_verifies_bob =
+        alice.encryption().get_verification(bob.user_id().unwrap(), "flow_id").await;
+    assert!(alice_verifies_bob.is_none());
+
+    let alice_verifies_bob_request =
+        alice.encryption().get_verification_request(&bob_user_id, "flow_id").await;
+    assert!(alice_verifies_bob_request.is_none());
+
+    let alice_bob_device =
+        alice.encryption().get_device(&bob_user_id, &bob_device_id).await.unwrap();
+    assert!(alice_bob_device.is_none());
+
+    // Have both Alice and Bob bootstrap cross-signing.
+    bootstrap_cross_signing(&alice).await;
+    bootstrap_cross_signing(&bob).await;
+
+    // Have Alice and Bob upload their signed device keys.
+    {
+        let mut sync_response_builder = SyncResponseBuilder::new();
+        mock_sync(&server.server, sync_response_builder.build_json_sync_response(), None).await;
+        alice.sync_once(Default::default()).await.unwrap();
+        bob.sync_once(Default::default()).await.unwrap();
+    }
+
+    // Have Alice track Bob (and thus fetches his identity).
+    {
+        let alice_olm = alice.olm_machine_for_testing().await;
+        let alice_olm = alice_olm.as_ref().unwrap();
+        alice_olm.update_tracked_users([bob_user_id.as_ref()]).await.unwrap();
+    }
+
+    // Run a sync so we do send outgoing requests, including the /keys/query for
+    // getting bob's identity.
+    let mut sync_response_builder = SyncResponseBuilder::new();
+    mock_sync(&server.server, sync_response_builder.build_json_sync_response(), None).await;
+    alice.sync_once(Default::default()).await.unwrap();
+
+    // From the point of view of Alice, Bob now has a device.
+    let alice_bob_device = alice
+        .encryption()
+        .get_device(&bob_user_id, &bob_device_id)
+        .await
+        .unwrap()
+        .expect("alice sees bob's device");
+
+    assert!(!alice_bob_device.is_verified());
+    assert!(!alice_bob_device.is_deleted());
+    assert!(alice_bob_device.verify().await.is_err(), "can't sign the device of another user");
+
+    let alice_bob_ident = alice
+        .encryption()
+        .get_user_identity(&bob_user_id)
+        .await
+        .unwrap()
+        .expect("alice sees bob's identity");
+
+    alice_bob_ident.verify().await.unwrap();
+
+    assert_eq!(alice_bob_ident.user_id(), bob_user_id);
+    assert!(!alice_bob_ident.is_verified()); // FIXME: this should be verified after manual verification??
```

Comment:
@poljar Is this a bug?

Reply:
No, the method uploads a signature and a `/keys/query` needs to be received from the server which adds the signature to the cross-signing keys.

We could opportunistically add the signature to the identity ourselves but as of now we require the remote echo to be received.